components/movementsensor/gpsnmea/pmtkI2C.go

```diff
@@ -7,360 +7,226 @@ import (
+		defer close(dr.data)
+
 		strBuf := ""
 		for {
 			select {
-			case <-g.cancelCtx.Done():
+			case <-dr.cancelCtx.Done():
 				return
 			default:
 			}
 
-			// Opening an i2c handle blocks the whole bus, so we open/close each loop so other things also have a chance to use it
-			handle, err := g.bus.OpenHandle(g.addr)
-			// Record the error value no matter what. If it's nil, this will help suppress
-			// ephemeral errors later.
-			g.err.Set(err)
-			if err != nil {
-				g.logger.CErrorf(ctx, "can't open gps i2c handle: %s", err)
-				return
-			}
-			buffer, err := handle.Read(ctx, 1024)
-			g.err.Set(err)
-			hErr := handle.Close()
-			g.err.Set(hErr)
-			if hErr != nil {
-				g.logger.CErrorf(ctx, "failed to close handle: %s", hErr)
-				return
-			}
+			buffer, err := dr.readData()
 			if err != nil {
-				g.logger.CError(ctx, err)
+				dr.logger.CErrorf(dr.cancelCtx, "failed to read data, retrying: %s", err)
 				continue
 			}
+
 			for _, b := range buffer {
 				// PMTK uses CRLF line endings to terminate sentences, but just LF to blank data.
-				// Since CR should never appear except at the end of our sentence, we use that to determine sentence end.
-				// LF is merely ignored.
-				if b == 0x0D {
+				// Since CR should never appear except at the end of our sentence, we use that to
+				// determine sentence end. LF is merely ignored.
+				if b == 0x0D { // 0x0D is the ASCII value for a carriage return
 					if strBuf != "" {
-						// sometimes we miss "$" on the first message of the buffer. Here we are adding the missing
-						// "$" to a valid nmea string.
-						if strBuf[0] == 0x47 { // 0x47 is the ASCII value
+						// Sometimes we miss "$" on the first message of the buffer. If the first
+						// character we read is a "G", it's likely that this has occurred, and we
+						// should add a "$" at the beginning.
+						if strBuf[0] == 0x47 { // 0x47 is the ASCII value for "G"
 							strBuf = "$" + strBuf
 						}
 
-						g.mu.Lock()
-						err = g.data.ParseAndUpdate(strBuf)
-						g.mu.Unlock()
-						if err != nil {
-							g.logger.CDebugf(ctx, "can't parse nmea sentence: %s, %v", strBuf, err)
-							g.logger.Debug("Check: GPS requires clear sky view." +
-								" Ensure the antenna is outdoors if signal is weak or unavailable indoors.")
+						dr.data <- strBuf
+						strBuf = ""
+
+						// Check if we're supposed to shut down again. Perhaps we waited a long
+						// time to put data into the channel.
+						select {
+						case <-dr.cancelCtx.Done():
+							return
+						default:
 						}
 					}
-					strBuf = ""
-				} else if b != 0x0A && b < 0x7F { // adds only valid bytes
+				} else if b != 0x0A && b < 0x7F { // only append valid (printable) bytes
 					strBuf += string(b)
 				}
 			}
 		}
 	})
-
-	return g.err.Get()
 }
 
-// GetBusAddr returns the bus and address that takes in rtcm corrections.
-func (g *PmtkI2CNMEAMovementSensor) GetBusAddr() (buses.I2C, byte) {
-	return g.bus, g.addr
```

Comment:
Just calling out this function n longer exists, but I think it's okay. I don't see anything using it in the rtk version.

Reply:
Yeah, I don't know why this function exists (there is no reason to make the bus public), and it's not used anywhere else in the RDK (so the only place it could be used is in a module or other third-party code, and I don't mind if we break something there).